examples/pytorch/fsdp/fine-tune-t5-with-fsdp.ipynb

```diff
@@ -0,0 +1,509 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This Notebook will fine-tune Text-to-Text Transfer Transformer (T5) with Wikihow dataset for text summarization using Kubeflow PyTorchJob.\n",
+    "\n",
+    "Pretrained T5 model: https://huggingface.co/google-t5/t5-base\n",
+    "\n",
+    "Wikihow dataset: https://github.com/mahnazkoupaee/WikiHow-Dataset\n",
+    "\n",
+    "This Notebook will use **4** GPUs to fine-tune T5 model on 2 Nodes and 4 GPUs. This example is based on [the official PyTorch FSDP tutorial](https://pytorch.org/tutorials/intermediate/FSDP_adavnced_tutorial.html)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## FSDP with multi-node multi-worker training"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This Notebook demonstrates multi-node, multi-worker distributed training with Fully Sharded Data Parallel (FSDP) and PyTorchJob.\n",
+    "\n",
+    "When a model is trained with FSDP, the GPU memory footprint is smaller compare to Distributed Data Parallel (DDP),\n",
+    "as the model parameters are sharded across GPU devices.\n",
+    "\n",
+    "This enables training of very large models that would otherwise be impossible to fit on a single GPU device.\n",
+    "\n",
+    "Check this guide to learn more about PyTorch FSDP: https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Install the required packages\n",
+    "\n",
+    "Install the Kubeflow Training Python SDK."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# TODO (andreyvelich): Use the release version of SDK.\n",
+    "!pip install git+https://github.com/kubeflow/training-operator.git#subdirectory=sdk/python"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create script to fine-tune T5 using FSDP\n",
+    "\n",
+    "We need to wrap our fine-tuning script in a function to create Kubeflow PyTorchJob."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def train_function(parameters):\n",
+    "    import os\n",
+    "    import time\n",
+    "    import functools\n",
+    "\n",
+    "    import torch\n",
+    "    import torch.distributed as dist\n",
+    "    from torch.utils.data.distributed import DistributedSampler\n",
+    "    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP\n",
+    "    from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy\n",
+    "\n",
+    "    from transformers import T5Tokenizer, T5ForConditionalGeneration\n",
+    "    from transformers.models.t5.modeling_t5 import T5Block\n",
+    "    from datasets import Dataset\n",
```

Comment:
It's not a problem, just curious that why prefer encapsule the whole script as a function and include import in it.

Reply:
This is how `create_job()` API works today with train function.
As you can see, we wrap the whole function body under containers args: https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/training/utils/utils.py#L134-L140, similar to KFP lightweight Python components.

Do you have any suggestions/ideas on how we can serialize function without asking users to make imports defined under the function ?

Reply:
Got it. Thanks for your information.
I have no better solution in mind now for this kind of usage. While I was thinking something like `function level scheduling` thing, despite KFP lightweight Python Components and Ray, please share more if you are familiar with other good practices.

Reply:
@kuizhiqing One of the options could be to use cloudpickle: https://github.com/cloudpipe/cloudpickle
That allows us to serialize function to byte string and then de-serialize it on the container.
That supports imports outside of serialized function.